pkg/client/apiutil/lazyrestmapper/lazyrestmapper_test.go

```diff
@@ -0,0 +1,228 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lazyrestmapper
+
+import (
+	"net/http"
+	"testing"
+
+	_ "github.com/onsi/ginkgo/v2"
+	gmg "github.com/onsi/gomega"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+)
+
+// countingRoundTripper is used to count HTTP requests.
+type countingRoundTripper struct {
+	roundTripper http.RoundTripper
+	requestCount int
+}
+
+func newCountingRoundTripper(rt http.RoundTripper) *countingRoundTripper {
+	return &countingRoundTripper{roundTripper: rt}
+}
+
+// RoundTrip implements http.RoundTripper.RoundTrip that additionally counts requests.
+func (crt *countingRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	crt.requestCount++
+	return crt.roundTripper.RoundTrip(r)
+}
+
+// GetRequestCount returns how many requests have been made.
+func (crt *countingRoundTripper) GetRequestCount() int {
+	return crt.requestCount
+}
+
+func setupEnvtest(t *testing.T) (*rest.Config, func(t *testing.T)) {
+	t.Log("Setup envtest")
+	g := gmg.NewWithT(t)
+	testEnv := &envtest.Environment{}
+	cfg, err := testEnv.Start()
+	g.Expect(err).NotTo(gmg.HaveOccurred())
+	g.Expect(cfg).NotTo(gmg.BeNil())
+
+	teardownFunc := func(t *testing.T) {
+		t.Log("Stop envtest")
+		g.Expect(testEnv.Stop()).To(gmg.Succeed())
+	}
+	return cfg, teardownFunc
+}
+
+func TestLazyRestMapperProvider(t *testing.T) {
+	restCfg, tearDownFn := setupEnvtest(t)
+	defer tearDownFn(t)
+
+	t.Run("getFilteredAPIGroupResources should return APIGroupResources based on passed arguments", func(t *testing.T) {
+		g := gmg.NewWithT(t)
+
+		httpClient, err := rest.HTTPClientFor(restCfg)
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+
+		crt := newCountingRoundTripper(httpClient.Transport)
+		httpClient.Transport = crt
+
+		discoveryClient, err := discovery.NewDiscoveryClientForConfigAndClient(restCfg, httpClient)
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+
+		// Get GroupResources for kubernetes core and kubernetes apps groups
+		filteredAPIGroupResources, err := getFilteredAPIGroupResources(discoveryClient, "apps")
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+		g.Expect(filteredAPIGroupResources.Group.Name).To(gmg.Equal("apps"))
+
+		// We expect to perform 3 requests to fetch apps group resources:
+		// #1: GET https://host/api
+		// #2: GET https://host/apis
+		// #3: GET https://host/apis/apps/v1
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(3))
+	})
+
+	t.Run("LazyRESTMapper should fetch data based on the request", func(t *testing.T) {
+		g := gmg.NewWithT(t)
+
+		// For each new group mapper performs 3 requests to the API server:
+		// #1: GET https://host/api
+		// #2: GET https://host/apis
+		// #3: GET https://host/apis/<group>/<version>
```

Comment:
Couple of questions:
1. This isn't actually tested (The test for `getFilteredAPIGroupResources` is insufficient, my reading of the discovery restmapper is that it will do these again) 
2. I don't understand why we need to do step 1. and 2. We know both the group and the version, so only step 3 should be enough
3. If I understand it correctly, as soon as we discover a second group we end up re-doing all these api calls plus the discovery call needed for the second group, rather than just the discovery call for the second group.

Given that this is a stop-gap change until aggregated discovery is available for everyone, I do not want to merge something that doesn't fully solve the original problem, it'll just lead to a new issue of `Provide a truly, truly lazy restmapper`

Reply:
Hey! Thank you for reviewing the code. I've optimized the mapper significantly, so now it is `The Truly Lazy Restmapper`.

Now, it needs 2 api calls (/api and /apis) to initialize + just one call for each requested api group version. I've added more tests to validate this behavior + refactored the code to make it simpler.

Let me know if there is something else I could improve.